File: lib/tools/account.ts
import { Currency, ExchangeRates } from "./forex";

export interface AccountInfo {
  places: number;
  currency: Currency;
  exchangeRates: ExchangeRates;
  amount: number;
  marginRisk: number;
  positionRisk: number;
}

export interface SetPlacesAccountAction {
  action: "setPlaces";
  places: number;
}

export interface SetCurrencyAccountAction {
  action: "setCurrency";
  currency: Currency;
}

export interface SetExchangeRatesAccountAction {
  action: "setExchangeRates";
  exchangeRates: ExchangeRates;
}

export interface SetAmountAccountAction {
  action: "setAmount";
  amount: number;
}

export interface SetMarginRiskAction {
  action: "setMarginRisk";
  risk: number;
}

export interface SetPositionRiskAction {
  action: "setPositionRisk";
  risk: number;
}

export type AccountAction =
  | SetPlacesAccountAction
  | SetCurrencyAccountAction
  | SetExchangeRatesAccountAction
  | SetAmountAccountAction
  | SetMarginRiskAction
  | SetPositionRiskAction;

export function accountReducer(
  state: AccountInfo,
  action: AccountAction
): AccountInfo {
  switch (action.action) {
    case "setPlaces":
      return { ...state, places: action.places };
    case "setCurrency":
      return { ...state, currency: action.currency };
    case "setExchangeRates":
      return { ...state, exchangeRates: action.exchangeRates };
    case "setAmount":
      return { ...state, amount: action.amount };
    case "setMarginRisk":
      return { ...state, marginRisk: action.risk };
    case "setPositionRisk":
      return { ...state, positionRisk: action.risk };
    default:
      console.error("Unrecognized account action", action);
      return state;
  }
}

File: src/model/trading/account.rs
use std::rc::Rc;

use gloo::storage::{errors::StorageError, Storage};
use serde::{Deserialize, Serialize};
use yew::Reducible;

use crate::model::currency::{Currency, ExchangeRates};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Account {
    pub places: usize,
    pub currency: Currency,
    pub exchange_rates: ExchangeRates,
    pub amount: f64,
    pub margin_risk: f64,
    pub position_risk: f64,
}

impl Default for Account {
    fn default() -> Self {
        let currency = Currency::GBP;

        Self {
            places: 4,
            currency,
            exchange_rates: ExchangeRates::new(currency),
            amount: 500.0,
            margin_risk: 0.01,
            position_risk: 0.01,
        }
    }
}

pub enum AccountAction {
    Load,
    SetPlaces { places: usize },
    SetCurrency { currency: Currency },
    SetExchangeRates { exchange_rates: ExchangeRates },
    SetAmount { amount: f64 },
    SetMarginRisk { risk: f64 },
    SetPositionRisk { risk: f64 },
}

impl Reducible for Account {
    type Action = AccountAction;

    fn reduce(self: Rc<Self>, action: Self::Action) -> Rc<Self> {
        let account = match action {
            AccountAction::Load => Self::load(),

            AccountAction::SetPlaces { places } => Self {
                places,
                ..(*self).clone()
            },

            AccountAction::SetCurrency { currency } => Self {
                currency,
                ..(*self).clone()
            },

            AccountAction::SetExchangeRates { exchange_rates } => Self {
                exchange_rates,
                ..(*self).clone()
            },

            AccountAction::SetAmount { amount } => Self {
                amount,
                ..(*self).clone()
            },

            AccountAction::SetMarginRisk { risk } => Self {
                margin_risk: risk,
                ..(*self).clone()
            },

            AccountAction::SetPositionRisk { risk } => Self {
                position_risk: risk,
                ..(*self).clone()
            },
        };

        account.save();
        account.into()
    }
}

impl Account {
    pub fn load() -> Self {
        match gloo::storage::LocalStorage::get("trading.account") {
            Ok(stored) => stored,
            Err(err) => match err {
                StorageError::KeyNotFound(_) => {
                    log::info!("No stored trading account found, using defaults");
                    Self::default()
                }

                _ => {
                    log::error!("Failed to retrieve trading account from local storage: {err:?}");
                    Self::default()
                }
            },
        }
    }

    pub fn save(&self) {
        if let Err(err) = gloo::storage::LocalStorage::set("trading.account", self) {
            log::error!("Failed to store trading account in local storage: {err:?}");
        }
    }
}
